### PR TITLE
Adapt for pytorch version >= 1.9

### PR DESCRIPTION
**Pull request by ksellesk**

### CODE REVIEW

**Comment by robin1001** [comment]
Can you describe the problem when you use the old code with Pytorch 1.9

**Comment by ksellesk** [comment]
i tried a lot, but all fails.

then i chage my code like this sample:

https://github.com/pytorch/examples/blob/master/imagenet/main.py


and it works!

**Comment by robin1001** [comment]
what's the error message?

**Comment by ksellesk** [comment]
Traceback (most recent call last):
  File "wenet/bin/train.py", line 199, in <module>
Traceback (most recent call last):
  File "wenet/bin/train.py", line 199, in <module>
    model, find_unused_parameters=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in __init__
    model, find_unused_parameters=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in __init__
    dist._verify_model_across_ranks(self.process_group, parameters)
    dist._verify_model_across_ranks(self.process_group, parameters)
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandled cuda error, NCCL version 2.7.8
ncclUnhandledCudaError: Call to CUDA function failed.RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandl
ed cuda error, NCCL version 2.7.8
ncclUnhandledCudaError: Call to CUDA function failed.





our machine support nccl.

**Comment by ksellesk** [comment]
Any progress?

Has anyone tried orch version >= 1.9.

**Comment by TeaPoly** [comment]
> Any progress?
> 
> Has anyone tried orch version >= 1.9?

GPU 1080 ti and TITAN RTX  has no problem in PyTorch 1.9.0, but 2080Ti has the same problem.

File: wenet/bin/train.py [hunk 1]
```diff
@@ -194,10 +193,11 @@
     if distributed:
         assert (torch.cuda.is_available())
         # cuda model is required for nn.parallel.DistributedDataParallel
-        model.cuda()
+        torch.cuda.set_device(args.rank)
```
Comment (TeaPoly) [inline wenet/bin/train.py:196]: Maybe it should be `args.gpu`.

Comment (ksellesk) [reply wenet/bin/train.py:196]: Logically it should be args.gpu, but i failed with args.gpu but succeed with args.rank.....

Comment (ksellesk) [reply wenet/bin/train.py:196]: https://github.com/espnet/espnet/blob/master/espnet2/train/trainer.py#L229

for reference

**Comment by TeaPoly** [comment]
> Traceback (most recent call last):
> File "wenet/bin/train.py", line 199, in 
> Traceback (most recent call last):
> File "wenet/bin/train.py", line 199, in 
> model, find_unused_parameters=True)
> File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> model, find_unused_parameters=True)
> File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> dist._verify_model_across_ranks(self.process_group, parameters)
> dist._verify_model_across_ranks(self.process_group, parameters)
> RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandled cuda error, NCCL version 2.7.8
> ncclUnhandledCudaError: Call to CUDA function failed.RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandl
> ed cuda error, NCCL version 2.7.8
> ncclUnhandledCudaError: Call to CUDA function failed.
> 
> .
> .
> 
> our machine supports nccl.

What is the graphics card you are using?

**Comment by ksellesk** [comment]
> > Traceback (most recent call last):
> > File "wenet/bin/train.py", line 199, in
> > Traceback (most recent call last):
> > File "wenet/bin/train.py", line 199, in
> > model, find_unused_parameters=True)
> > File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> > model, find_unused_parameters=True)
> > File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> > dist._verify_model_across_ranks(self.process_group, parameters)
> > dist._verify_model_across_ranks(self.process_group, parameters)
> > RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandled cuda error, NCCL version 2.7.8
> > ncclUnhandledCudaError: Call to CUDA function failed.RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandl
> > ed cuda error, NCCL version 2.7.8
> > ncclUnhandledCudaError: Call to CUDA function failed.
> > .
> > .
> > our machine supports nccl.
> 
> What is the graphics card you are using?


3090

**Comment by shanguanma** [comment]
> > > Traceback (most recent call last):
> > > File "wenet/bin/train.py", line 199, in
> > > Traceback (most recent call last):
> > > File "wenet/bin/train.py", line 199, in
> > > model, find_unused_parameters=True)
> > > File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> > > model, find_unused_parameters=True)
> > > File "/opt/conda/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 496, in **init**
> > > dist._verify_model_across_ranks(self.process_group, parameters)
> > > dist._verify_model_across_ranks(self.process_group, parameters)
> > > RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandled cuda error, NCCL version 2.7.8
> > > ncclUnhandledCudaError: Call to CUDA function failed.RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1623448265233/work/torch/lib/c10d/ProcessGroupNCCL.cpp:911, unhandl
> > > ed cuda error, NCCL version 2.7.8
> > > ncclUnhandledCudaError: Call to CUDA function failed.
> > > .
> > > .
> > > our machine supports nccl.
> > 
> > 
> > What is the graphics card you are using?
> 
> 3090

I occurs same error on the nvidia RTX 3080

**Comment by robin1001** [comment]
any update now?

**Comment by shanguanma** [comment]
Util now, I found that one way was work for me in the pytorch1.8 or pytorch1.9, 
I use the command `docker run -it --rm --gpus '"device=1,2"' --shm-size=1g --ulimit memlock=-1 --ipc=host  -v /mnt/4T:/mnt/4T -v /home/maduo/docker_asr_wenet_nvidia_gpu/wenet:/code/w2021/wenet asr_wenet_nvidia_gpu:latest` to run wenet,  but in the run.sh I must be set ``export CUDA_VISIBLE_DEVICES="0,1"  If I increase number of  gpus, it will occur the above error.

**Comment by ksellesk** [comment]
Any progress?

File: wenet/bin/train.py (hunk 1 above)
Comment (robin1001) [reply wenet/bin/train.py:196]: what if we use multiple GPUs on multiple machine?

**Comment by cnrpman** [comment]
> Any progress?

Seems the fix is working weird when loading checkpoints.. I attached my details in https://github.com/wenet-e2e/wenet/discussions/586#discussioncomment-1345310

**Pull request closed without merging**